lib/active_model/serializer/association.rb

```diff
@@ -9,10 +9,46 @@ class Serializer
     # @example
     #  Association.new(:comments, CommentSummarySerializer)
     #
-    Association = Struct.new(:name, :serializer, :options, :links, :meta) do
+    Association = Struct.new(:name, :serializer, :options) do
+      attr_reader :reflection_options
+
+      def initialize(*)
+        super
+        @reflection_options = options.delete(:new_options)
+      end
+
       # @return [Symbol]
       def key
-        options.fetch(:key, name)
+        @key ||= options.fetch(:key) { name }
+      end
+
+      def links
+        reflection_options.fetch(:links)
+      end
+
+      def meta
+        reflection_options.fetch(:meta)
+      end
+
+      def reflection
+        reflection_options.fetch(:reflection)
+      end
+
+      def to_many?
+        reflection.name == :has_many
+      end
+
+      def belongs_to?
+        reflection.foreign_key_on == :self
+      end
+
+      def serializer
```

Comment:
the serializer really shouldn't be evaluated until needed. Unfortunately, a lot of our code very much depends on expecting it to be evaluated. This implementation makes it possible to make it lazy.  It's not ideal, but is just a step to getting it right.

Unfortunately, associations are evaluated in two places: in `Serializer#relationship_value_for` and in `JsonApi#relationships_for`. Would be easier in one spot.  

The other problem, is that associations are built using the include directive, but it's not clear when the attributes are requested, especially since the include directive relies on the `*` behavior

You'd think

```ruby
        requested_associations = fieldset.fields_for(resource_object[:type]) || '*'
        relationships = relationships_for(serializer, requested_associations)
```

would imply which fields to get....